File: 프로그래머스/lv2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.cpp
#include <string>
#include <vector>
#include <cctype>
#include <tuple>
#include <algorithm>
#include <cstring>

using namespace std;
using fname = tuple<string, string, string>;

vector<string> solution(vector<string> files) {
    string HEAD, NUMBER, TAIL;
    vector<fname> v;
    for(const string &file : files)
    {
        int i = 0, j = 0;
        
        while(!isdigit(file[j])) j++;
        HEAD = string(&file[i], &file[j]);
        i = j;
        
        while(isdigit(file[j])) j++;
        NUMBER = string(&file[i], &file[j]);
        i = j;
        
        TAIL = file.substr(i);
        
        v.emplace_back(HEAD, NUMBER, TAIL);
    }
    
    stable_sort(v.begin(), v.end(), [](const fname &a, const fname &b) {
        const auto &[h1, n1, t1] = a;
        const auto &[h2, n2, t2] = b;
        
        int result = strcasecmp(h1.c_str(), h2.c_str());
        if(result == 0) return stoi(n1) < stoi(n2);
        return result < 0;
    });
    
    vector<string> answer;
    for(const auto &[head, number, tail] : v)
    {
        answer.emplace_back(head + number + tail);
    }
    return answer;
}

File: 프로그래머스/lv2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.py
import re

def divide(x):
    m = re.match('([a-zA-Z \.-]+)([0-9]{1,5})(.*)', x)
    head, number, tail = m.groups()
    return head, number, tail

def solution(files):
    for k, file in enumerate(files):
        files[k] = [file, divide(file)]
        
    answer = sorted(files, key = lambda x: (
        x[1][0].lower(),
        int(x[1][1])
    ))
    
    for k, v in enumerate(answer):
        answer[k] = v[0]
        
    return answer